apollo-client/src/main/java/com/ctrip/framework/apollo/spring/property/SpringCloudAutoUpdateConfigChangeListener.java

```diff
@@ -0,0 +1,50 @@
+/*
+ * Copyright 2022 Apollo Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+package com.ctrip.framework.apollo.spring.property;
+
+import com.ctrip.framework.apollo.model.ConfigChangeEvent;
+import org.springframework.beans.factory.config.ConfigurableListableBeanFactory;
+import org.springframework.cloud.context.environment.EnvironmentChangeEvent;
+import org.springframework.cloud.context.scope.refresh.RefreshScope;
+import org.springframework.context.ApplicationContext;
+import org.springframework.core.env.Environment;
+
+public class SpringCloudAutoUpdateConfigChangeListener extends AutoUpdateConfigChangeListener {
+  final ApplicationContext applicationContext;
+
+  public SpringCloudAutoUpdateConfigChangeListener(
+      Environment environment,
+      ConfigurableListableBeanFactory beanFactory,
+      ApplicationContext applicationContext) {
+    super(environment, beanFactory);
+    this.applicationContext = applicationContext;
+  }
+
+  @Override
+  public void onChange(ConfigChangeEvent changeEvent) {
+    super.onChange(changeEvent);
+    adapterSpringCloud(changeEvent);
+  }
+
+  private void adapterSpringCloud(ConfigChangeEvent changeEvent) {
+    RefreshScope refreshScope = applicationContext.getBean(RefreshScope.class);
+    if (refreshScope != null) {
+      refreshScope.refreshAll();
```

Comment:
`refreshAll` is kind of a dangerous operation, it could cause unexpected behaviors, e.g. eureka service up and down. So I think we need to allow users to customize the change listener behaviors.

There was a config `apollo.autoUpdateInjectedSpringProperties` to control whether AutoUpdateConfigChangeListener is enabled or not. I think we need to also have two configs to control whether `refreshScope.refreshAll` and `applicationContext.publishEvent` are enabled respectively.

With that said, I would suggest we refactor the change listener logic instead of adding a new SpringCloudConfigPropertySourceProcessor
1. Define a Service Provider Interface, e.g. SpringAutoUpdateConfigChangeListener
2. Implement different implementations: AutoUpdateConfigChangeListener(the current one), SpringCloudRefreshScopeConfigChangeListener, SpringCloudEnviromentChangeEventConfigChangeListener
3. Use ServiceBootstrap#loadAllOrdered method to load all the implementations in PropertySourcesProcessor#initializeAutoUpdatePropertiesFeature
4. Write a CompositeConfigChagneListener to trigger each of the change listener
5. In the onChange method, each listener will first check whether itself is enabled, then will do the actual work if enabled. 

![image](https://user-images.githubusercontent.com/837658/162347348-cb7ee27b-f01b-4d86-99ef-91153cd72d1c.png)

The benefit of this architecture is it's easy to add more customized config change listeners in the future.

If you like this idea, I think I could contribute to the refactoring part.

Reply:
Yeah, Sounds good, l prefer it.

Reply:
I submitted #4305 to enable the spi mechanism, please help to take a look.

Reply:
> I submitted #4305 to enable the spi mechanism, please help to take a look.

Got it.